File: 2015/ruby/3/main-part-1.rb
#!/usr/bin/env ruby

class Position 

    attr_accessor :x, :y

    def initialize(x, y)
        self.x = x
        self.y = y
    end

    def move(direction)
        case direction
        when '^'
            self.y += 1
        when 'v'
            self.y -= 1
        when '>'
            self.x += 1
        when '<'
            self.x -= 1
        end
    end
end

def main()
    args = ARGV

    if args.length < 1
        puts "Usage: ruby main.rb <input.txt>"
        exit 1
    end

    filename = args[0]
    file = File.open(filename, "r")

    if file.nil?
        puts "Could not open file"
        exit 1
    end

    lines = file.readlines
    file.close

    santa = Position.new(0, 0)
    grid = Hash.new(0)
    grid[[0, 0]] = 1

    lines.each do |line|
        line.each_char do |char|
            santa.move(char)
            grid[[santa.x, santa.y]] += 1
        end
    end

    puts "Part 1: #{grid.length}"
end

main()

File: 2015/python/3/main-part-1.py
#!/usr/bin/env python3

import sys
import os

class Position:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def move(self, direction):
        if direction == "^":
            self.y += 1
        elif direction == "v":
            self.y -= 1
        elif direction == "<":
            self.x -= 1
        elif direction == ">":
            self.x += 1

    
def main():
    # get command line arguments
    args = sys.argv[1:]
    if len(args) == 0:
        print("Usage: python main.py <input_file>")
        sys.exit(1)

    # get input file
    input_file = args[0]
    if not os.path.exists(input_file):
        print("File not found: {}".format(input_file))
        sys.exit(1)

    # read input file
    with open(input_file, "r") as f:
        lines = f.readlines()
        pos = Position(0, 0)

        grid = {}
        grid[(0, 0)] = 1

        for line in lines:
            for char in line:
                pos.move(char)
                if (pos.x, pos.y) in grid:
                    grid[(pos.x, pos.y)] += 1
                else:
                    grid[(pos.x, pos.y)] = 1

            print("Part 1:", len(grid))

if __name__ == "__main__":
    main()